oprf/oprf.go

```diff
@@ -0,0 +1,277 @@
+	OPRFP521 SuiteID = 0x0005
+)
+
+var (
+	// OPRFMode is the context string to define a OPRF.
+	OPRFMode byte = 0x00
+)
+
+var (
+	// ErrUnsupportedGroup is an error stating that the ciphersuite chosen is not supported
+	ErrUnsupportedGroup = errors.New("the chosen group is not supported")
+)
+
+// BlindToken corresponds to a token that has been blinded.
+// Internally, it is a serialized Element.
+type BlindToken []byte
+
+// IssuedToken corresponds to a token that has been issued.
+// Internally, it is a serialized Element.
+type IssuedToken []byte
+
+// Token is the object issuance of the protocol.
+type Token struct {
+	data  []byte
+	blind *group.Scalar
+}
+
+// Evaluation corresponds to the evaluation over a token.
+type Evaluation struct {
+	element []byte
+}
+
+// KeyPair is an struct containing a public and private key.
+type KeyPair struct {
+	pubK  *group.Element
+	PrivK *group.Scalar
+}
+
+// Client is a representation of a Client during protocol execution.
+type Client struct {
+	suite *group.Ciphersuite
+	ctx   []byte
+}
+
+// Server is a representation of a Server during protocol execution.
+type Server struct {
+	suite *group.Ciphersuite
+	ctx   []byte
+	Kp    *KeyPair
+}
+
+func generateCtx(id SuiteID) []byte {
+	ctx := [3]byte{OPRFMode, 0, byte(id)}
+
+	return ctx[:]
+}
+
+// Serialize serializes a KeyPair elements into byte arrays.
+func (kp *KeyPair) Serialize() ([]byte, []byte) {
+	pubK := kp.pubK.Serialize()
+	privK := kp.PrivK.Serialize()
+
+	return pubK, privK
+}
+
+// Deserialize deserializes a KeyPair into an element and field element of the group.
+func (kp *KeyPair) Deserialize(suite *group.Ciphersuite, privK, pubK []byte) error {
+	priv := group.NewScalar(suite)
+	priv.Deserialize(privK)
+
+	pub := group.NewElement(suite)
+	err := pub.Deserialize(pubK)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//GenerateKeyPair generates a KeyPair in accordance with the group.
+func GenerateKeyPair(suite *group.Ciphersuite) *KeyPair {
+	privK := suite.RandomScalar()
+	pubK := suite.ScalarMultBase(privK)
+
+	return &KeyPair{pubK, privK}
+}
+
+func assignKeyPair(suite *group.Ciphersuite, privK, pubK []byte) (*KeyPair, error) {
+	kp := &KeyPair{}
+	err := kp.Deserialize(suite, privK, pubK)
+	if err != nil {
+		return nil, err
+	}
+
+	return kp, nil
+}
+
+func suiteFromID(id SuiteID, ctx []byte) (*group.Ciphersuite, error) {
+	var err error
+	var suite *group.Ciphersuite
+
+	switch id {
+	case OPRFP256:
+		suite, err = group.NewSuite("P-256", ctx)
+	case OPRFP384:
+		suite, err = group.NewSuite("P-384", ctx)
+	case OPRFP521:
+		suite, err = group.NewSuite("P-521", ctx)
+	default:
+		return suite, ErrUnsupportedGroup
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return suite, err
+}
+
+// NewServer creates a new instantiation of a Server. It can create
+// a server with existing keys or use pre-generated keys.
+func NewServer(id SuiteID, privK, pubK []byte) (*Server, error) {
```

Comment:
Is it more idiomatic in Go to support optional parameters, or to have separate functions with different parameter lists? For example, rather than make these optional, we could have two functions like so:

```
func NewServer(id SuiteID) (*Server, error) // generate a key pair
func NewServerWithKeyPair(id SuiteID, privK, pubK []byte) (*Server, error) // assign the key pair
```

Reply:
Neither optional parameters (in its full PL definition) nor function overloading are supported in Go: https://golang.org/doc/faq#overloading, https://github.com/golang/go/wiki/GoForCPPProgrammers#top 
I could use this: Using Variadic Functions with Option Functions from https://medium.com/@petomalina/dealing-with-optional-parameters-in-go-9780f9bfbd1d

Reply:
I didn't overload the functions above -- I used separate names. I don't think variadic functions is the right answer here.

Reply:
Sure. Using that then.